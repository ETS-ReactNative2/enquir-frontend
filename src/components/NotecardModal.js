import React, { Component } from 'react';
import axios from 'axios';
import { Modal, TouchableOpacity, Text, KeyboardAvoidingView, Button, View, StyleSheet, TextInput, ActivityIndicator } from 'react-native';
import QuestionModal from './QuestionModal';
import Storage from '../storage/Storage';

const storage = new Storage();
const WEB_SCRAPER_API_URL = 'EC2Co-EcsEl-3QPW7ZVE3WXS-1593988200.us-west-1.elb.amazonaws.com:5000/autopopcontent';
const QUESTION_GENERATOR_API_URL = 'https://rocky-caverns-51964.herokuapp.com/genquest';

export default class NotecardModal extends Component {
    state = {
        isLoading: false,
        noteItem: null,
        showingQuestionModal: false,
        topicInput: '',
        notesInput: '',
        questions: [],
    };

    resetNotecard = () => {
        this.setState({
            isLoading: false,
            noteItem: null,
            topicInput: '',
            notesInput: '',
            questions: [],
        });
    };

    // Called by Main component to pass data to modal
    loadData = (item) => {
        // If no item is passed, reset state of modal to represent an empty notecard,
        // otherwise populate the notecard with info
        if (item == undefined) {
            this.resetNotecard();
        } else {
            this.setState({
                noteItem: item,
                topicInput: item.topic,
                notesInput: item.notes,
                questions: item.questions,
            });
        }
    };

    // Close the modal in parent component
    onClose = () => {
        // TODO: modal asking 'Are you sure? Unsaved changes will be lost!'
        this.props.onClose && this.props.onClose();
    };

    // Handles opening of question modal
    showQuestionModal = () => {
        this.setState({
            showingQuestionModal: !this.state.showingQuestionModal
        });
    };

    // Saves notecard and generates questions
    onSave = async () => {
        this.setState({
            isLoading: true,    // show loader while request is performing
        });

        // get existing notes
        const existingNotes = await storage.retrieveNotes();

        // generate questions
        await this.generateQuestions();

        // default topic name
        const topic = this.state.topicInput.length > 0 ? this.state.topicInput : 'Untitled';

        // overwrite notecard
        if (this.state.noteItem) {
            const index = (existingNotes.findIndex((note) => note.id == this.state.noteItem.id)).toString();
            existingNotes[index] = {
                id: index,
                topic: topic,
                notes: this.state.notesInput,
                questions: this.state.questions,
            }
        } else {
            // add new notecard
            const lastIndex = existingNotes.length > 0 ? existingNotes[existingNotes.length - 1].id : -1;
            const newIndex = (Number.parseInt(lastIndex) + 1).toString();

            existingNotes.push({
                id: newIndex,
                topic: topic,
                notes: this.state.notesInput,
                questions: this.state.questions,
            });
        }

        // write to storage
        await storage.addNotes(existingNotes);

        // reset modal
        this.resetNotecard();

        this.props.onSave && this.props.onSave();
    };

    // Scrape web page for content on a topic
    autoPop = async () => {
        try {
            this.setState({
                isLoading: true,    // show loader while request is performing
            });

            const res = await axios.post(WEB_SCRAPER_API_URL, {
                query: this.state.topicInput,
            });

            this.setState({
                isLoading: false,
                notesInput: res,
            });
        } catch (err) {
            console.log('Err occurred communicating with scraper: ', err);
        }
    };

    // Generate questions given content
    generateQuestions = async () => {
        try {
            const res = await axios.post(QUESTION_GENERATOR_API_URL, {
                blurb: this.state.notesInput,
            });

            this.setState({
                questions: res.data.questions,
            });
        } catch (err) {
            console.log('Err occurred communicating with question generator: ', err);
        }
    };

    render() {
        if (!this.props.show) {
            return null;
        }

        const { isLoading, showingQuestionModal, notesInput, questions } = this.state;

        return (
            <Modal
                animationType='slide'
                onRequestClose={this.onClose} >
                <QuestionModal
                    show={showingQuestionModal}
                    onClose={this.showQuestionModal}
                    questions={questions}
                />
                {
                    isLoading &&
                    <ActivityIndicator size='large' color='#0064e1' />
                }

                {
                    !isLoading &&
                    <View style={{ flex: 1 }}>
                        <View style={styles.menu}>
                            <View style={styles.closeContainer}>
                                <TouchableOpacity
                                    style={styles.closeButton}
                                    onPress={this.onClose} >
                                    <Text style={styles.closeButtonText}>{'<'}</Text>
                                </TouchableOpacity>
                            </View>
                            <View style={styles.questionContainer}>
                                <TouchableOpacity
                                    syle={styles.questionButton}
                                    onPress={this.showQuestionModal} >
                                    <Text style={styles.questionButtonText}>?</Text>
                                </TouchableOpacity>
                            </View>
                            <View style={styles.saveContainer}>
                                <Button
                                    syle={styles.saveButton}
                                    onPress={this.onSave}
                                    title='Save'
                                />
                            </View>
                        </View>
                        <KeyboardAvoidingView
                            style={{ flex: 1 }}
                            behavior="padding"
                            enabled >
                            <View style={styles.container}>
                                <View style={styles.topicContainer}>
                                    <TextInput
                                        style={styles.topicInput}
                                        onChangeText={(topicInput) => this.setState({ topicInput })}
                                        value={this.state.topicInput}
                                        placeholder={'Enter topic'}
                                    />
                                </View>
                                <View style={styles.autoPopContainer}>
                                    <Button
                                        style={styles.autoPopButton}
                                        onPress={this.autoPop}
                                        title='auto'
                                    />
                                </View>
                                <View style={styles.notesContainer}>
                                    <TextInput
                                        style={styles.notesInput}
                                        multiline={true}
                                        onChangeText={(notesInput) => this.setState({ notesInput })}
                                        value={notesInput}
                                        placeholder={'Enter notes'}
                                        textAlignVertical={'top'}
                                    />
                                </View>
                            </View>
                        </KeyboardAvoidingView>
                    </View>
                }
            </Modal>
        );
    }
};

const styles = StyleSheet.create({
    menu: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    closeContainer: {
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'flex-start',
    },
    closeButton: {
        width: 65,
        height: 40,
        backgroundColor: 'white',
        justifyContent: 'center',
    },
    closeButtonText: {
        fontSize: 50,
        alignSelf: 'center',
        color: 'grey',
    },
    questionContainer: {
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'center',
    },
    questionButton: {
        // overflow: 'hidden',
        width: 60,
        height: 60,
        position: 'absolute',
        // left: 80,
        // bottom: 62,
        // borderTopLeftRadius: 150,
        // borderBottomLeftRadius: 150,
        backgroundColor: 'black'
    },
    questionButtonText: {
        fontSize: 24,
    },
    saveContainer: {
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'flex-end',
    },
    saveButton: {
    },
    container: {
        flex: 1,
    },
    topicContainer: {
        flexDirection: 'row',
        backgroundColor: "#eaeaea",
        marginHorizontal: 25,
        marginTop: 20,
        marginBottom: 10,
    },
    topicInput: {
        flex: 1,
        borderColor: "#ccc",
        margin: 10,
        // fontFamily: 'Roboto',
        fontSize: 19,
    },
    autoPopContainer: {
        flexDirection: 'row',
        justifyContent: 'flex-end',
        marginRight: 20,
    },
    autoPopButton: {
    },
    notesContainer: {
        flex: 2,
        flexDirection: 'row',
        backgroundColor: "#eaeaea",
        borderRadius: 4,
        marginHorizontal: 25,
        marginTop: 10,
        marginBottom: 35,
        elevation: 10,
    },
    notesInput: {
        flex: 1,
        margin: 10,
        // fontFamily: 'Roboto',
        fontSize: 18,
    }
});
