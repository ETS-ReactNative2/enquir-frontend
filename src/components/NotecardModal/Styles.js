import { StyleSheet } from 'react-native';

const containers = StyleSheet.create({
  menu: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 5,
  },
  closeContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
    width: 65,
    height: 30,
    marginTop: 5,
    backgroundColor: 'white',
  },
  saveContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
  content: {
    flex: 1,
  },
  topicContainer: {
    flexDirection: 'row',
    backgroundColor: "#eaeaea",
    marginHorizontal: 25,
    marginBottom: 10,
  },
  tabContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginRight: 20,
  },
  notesContainer: {
    flex: 2,
    flexDirection: 'row',
    backgroundColor: "#eaeaea",
    borderRadius: 4,
    marginHorizontal: 25,
    marginBottom: 35,
    elevation: 10,
  },
  questionsHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 15,
    elevation: 5,
    backgroundColor: '#d1d1d1',
  },
  emptyBlock: {
    paddingVertical: 36,
    paddingLeft: 18,
    paddingRight: 16,
    marginTop: 0,
    marginBottom: 8,
  },
});

const buttons = StyleSheet.create({
  closeButton: {
    marginLeft: 16,
    marginVertical: 7,
  },
  questionTab: {
    flexDirection: 'row',
    alignSelf: 'flex-end',
    justifyContent: 'center',
    width: 50,
    height: 45,
    marginTop: 5,
    marginRight: 6,
    borderTopLeftRadius: 30,
    borderTopRightRadius: 30,
    elevation: 10,
    backgroundColor: '#35477d',
  },
  autoPopButtonContainer: {
    flexDirection: 'row',
    alignSelf: 'flex-start',
    justifyContent: 'center',
    marginLeft: 37,
    marginBottom: 5,
    height: 40,
    width: 40,
    borderRadius: 40,
    backgroundColor: '#8ac6d1',
  },
  questionHeaderButtons: {
    flexDirection: 'row',
    justifyContent: 'flex-end'
  },
  questionGenButton: {},
  notifButton: {
    marginTop: 2.5,
    marginLeft: 15,
  },
  addQuestionButton: {
    width: 55,
    height: 55,
    borderRadius: 55 / 2,
    backgroundColor: '#00BCD4',
    position: 'absolute',
    bottom: 20,
    right: 20,
    alignSelf: 'flex-end',
    justifyContent: 'center',
    elevation: 7,
  },
});

const inputs = StyleSheet.create({
  topicInput: {
    flex: 1,
    borderColor: "#ccc",
    margin: 10,
    fontSize: 19,
    fontFamily: 'Roboto-Medium',
  },
  notesInput: {
    flex: 1,
    margin: 10,
    fontSize: 18,
    fontFamily: 'Roboto-Regular',
  },
});

const text = StyleSheet.create({
  saveButtonText: {
    fontSize: 20,
    fontFamily: 'Roboto-Medium',
    marginRight: 18,
  },
  questionHeaderText: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
    fontSize: 22,
    fontFamily: 'Roboto-Medium',
  },
});

export { containers, buttons, inputs, text };