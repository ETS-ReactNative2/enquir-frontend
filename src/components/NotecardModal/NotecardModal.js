import React, { Component } from 'react';
import axios from 'axios';
import {
  Alert,
  Modal,
  TouchableOpacity,
  TouchableWithoutFeedback,
  FlatList,
  Text,
  Keyboard,
  View,
  TextInput,
} from 'react-native';
import { containers, buttons, inputs, text } from './Styles';
import Icon from 'react-native-vector-icons/FontAwesome';
import Question from '../Question/Question';
import { Storage } from '../../services/Storage';
import { PushNotification } from '../../services/PushNotification';
import Loader from '../Loader/Loader';

const WEB_SCRAPER_API_URL = 'https://czx2q94gxb.execute-api.us-west-1.amazonaws.com/dev/autopopcontent';
const QUESTION_GENERATOR_API_URL = 'https://rocky-caverns-51964.herokuapp.com/genquest';

// Dismiss keyboard when touching anywhere
const DismissKeyboard = ({ children }) => (
  <TouchableWithoutFeedback onPress={() => Keyboard.dismiss()}>
    {children}
  </TouchableWithoutFeedback>
);

// For displaying tab icon
const TabIcon = ({ showingQuestions }) => {
  if (showingQuestions) {
    return <Icon name="pencil" size={35} color='#fffdf9' style={{ marginTop: 5 }} />;
  }
  return <Icon name="question" size={35} color='#fffdf9' style={{ marginTop: 5 }} />;
};

export default class NotecardModal extends Component {
  state = {
    isLoading: false,
    noteItem: null,
    showingQuestions: false,
    showingAddQuestionButton: true,
    topicInput: '',
    notesInput: '',
    questions: [],
  };

  resetNotecard = () => {
    this.setState({
      isLoading: false,
      noteItem: null,
      showingQuestions: false,
      showingAddQuestionButton: true,
      topicInput: '',
      notesInput: '',
      questions: [],
    });
  };

  // Called by Main component to pass data to modal
  loadData = (item) => {
    // If no item is passed, reset state of modal to represent an empty notecard,
    // otherwise populate the notecard with info
    if (item == null) {
      this.resetNotecard();
    } else {
      this.setState({
        noteItem: item,
        topicInput: item.topic,
        notesInput: item.notes,
        showingQuestions: false,
        questions: item.questions,
      });
    }
  };

  // Close the modal in parent component
  onClose = () => {
    // TODO: modal asking 'Are you sure? Unsaved changes will be lost!'
    this.props.onClose && this.props.onClose();
  };

  // Handles display of questions
  showQuestions = () => {
    this.setState({
      showingQuestions: !this.state.showingQuestions,
    });
  };

  // Saves notecard and generates questions
  onSave = async () => {
    // PushNotification.localNotification('Test Topic', 'When was I born?');
    PushNotification.localNotificationSchedule('Schedule Test', 'Where am I?');
    this.setState({
      isLoading: true,    // show loader while request is performing
      loadingState: 'Saving...',
    });

    // get existing notes
    const existingNotes = await Storage.retrieveNotes();

    // default topic name
    const topic = this.state.topicInput.length > 0 ? this.state.topicInput : 'Untitled';

    // overwrite notecard
    if (this.state.noteItem) {
      const index = existingNotes.findIndex((note) => note.id == this.state.noteItem.id);
      existingNotes[index] = {
        id: index,
        topic: topic,
        notes: this.state.notesInput,
        questions: this.state.questions,
      }
    } else {
      // add new notecard
      const newIndex = existingNotes.length == 0 ? 0 : existingNotes[existingNotes.length - 1].id + 1;

      existingNotes.push({
        id: newIndex,
        topic: topic,
        notes: this.state.notesInput,
        questions: this.state.questions,
      });
    }

    // write to storage
    await Storage.addNotes(existingNotes);

    // reset modal
    this.resetNotecard();

    this.props.onSave && this.props.onSave();
  };

  // Scrape web page for content on a topic
  autoPop = async () => {
    const { topicInput, notesInput } = this.state;
    if (topicInput.length < 1) {
      Alert.alert('To use the auto notecard maker, a topic must be provided');
      return;
    }

    try {
      this.setState({
        isLoading: true,    // show loader while request is performing
        loadingState: 'Gathering notes...',
      });

      const res = await axios.post(WEB_SCRAPER_API_URL, {
        query: this.state.topicInput,
      });

      if (res.status !== 200) {
        throw Error;
      }

      const { data } = res;

      // Add a couple newlines if notes already exist
      this.setState({
        notesInput: notesInput.length ? notesInput.concat(`\n\n${data.blurb}`) : data.blurb,
      });
    } catch (err) {
      console.log('Err occurred communicating with scraper: ', err);
      Alert.alert('Sorry!', `Couldn\'t find notes for "${topicInput}"`);
    }
    this.setState({
      isLoading: false,
    });
  };

  // Generate questions given content
  generateQuestions = async () => {
    try {
      this.setState({
        isLoading: true,    // show loader while request is performing
        loadingState: 'Creating questions...',
      });

      const res = await axios.post(QUESTION_GENERATOR_API_URL, {
        blurb: this.state.notesInput,
      });

      const { questions } = res.data;

      if (questions.length < 1) {
        Alert.alert('Sorry!', 'Could not generate questions for this notecard');
      } else {
        this.setState({
          questions: this.state.questions.concat(res.data.questions),    // add to existing questions
        });
      }
    } catch (err) {
      console.log('Err occurred communicating with question generator: ', err);
      Alert.alert('ERROR', 'Unable to generate questions');
    }
    this.setState({
      isLoading: false,
    });
  };

  // Add an empty question
  addQuestion = () => {
    const { questions } = this.state;
    const id = questions.length == 0 ? 0 : questions[questions.length - 1].id + 1;
    this.setState({
      questions: this.state.questions.concat({
        id,
        question: '',
      }),
    });
  };

  // Overwrite an existing questions content
  saveQuestion = (id, newQuestion) => {
    var questions = null;
    for (var i = 0; i < this.state.questions.length; ++i) {
      if (this.state.questions[i].id == id) {
        questions = [...this.state.questions];
        questions[i].question = newQuestion;
        break;
      }
    }

    if (questions == null) {
      console.log('Error occurred while trying to save question.');
      return;
    }

    this.setState({
      questions: questions,
    });
  };

  // Removes a question
  deleteQuestion = (id) => {
    var questions = null;
    for (var i = 0; i < this.state.questions.length; ++i) {
      if (this.state.questions[i].id == id) {
        var questions = [
          ...this.state.questions.slice(0, i),
          ...this.state.questions.slice(i + 1)
        ];
        break;
      }
    }

    if (questions == null) {
      console.log('Error occurred: could not find question to delete');
      return;
    }

    this.setState({
      questions: questions,
    });
  };

  // Keyboard listener for hiding/showing a button if keyboard is active/inactive
  _keyboardDidShow = () => {
    this.setState({
      showingAddQuestionButton: false,
    });
  }

  _keyboardDidHide = () => {
    this.setState({
      showingAddQuestionButton: true,
    });
  }

  componentDidMount() {
    this.keyboardDidShowListener = Keyboard.addListener('keyboardDidShow', this._keyboardDidShow);
    this.keyboardDidHideListener = Keyboard.addListener('keyboardDidHide', this._keyboardDidHide);
  };

  componentWillUnmount() {
    this.keyboardDidShowListener.remove();
    this.keyboardDidHideListener.remove();
  };

  render() {
    if (!this.props.show) {
      return null;
    }

    const { isLoading, loadingState, showingQuestions, showingAddQuestionButton, topicInput, notesInput, questions } = this.state;

    return (
      <Modal
        animationType='slide'
        onRequestClose={this.onClose} >
        <DismissKeyboard>
          <View style={{ flex: 1 }}>
            <Loader
              loading={isLoading}
              loadText={loadingState}
            />
            <View style={containers.menu}>
              <TouchableOpacity
                style={buttons.closeButton}
                onPress={this.onClose} >
                <View style={containers.closeContainer}>
                  <Icon name='chevron-left' size={30} />
                </View>
              </TouchableOpacity>
              <View style={containers.saveContainer}>
                <TouchableOpacity
                  onPress={this.onSave} >
                  <Text style={text.saveButtonText}>SAVE</Text>
                </TouchableOpacity>
              </View>
            </View>
            <View style={containers.content}>
              <View style={containers.topicContainer}>
                <TextInput
                  style={inputs.topicInput}
                  onChangeText={(topicInput) => this.setState({ topicInput })}
                  value={topicInput}
                  placeholder={'Enter topic'}
                />
              </View>
              <View style={containers.tabContainer}>
                <TouchableOpacity
                  onPress={this.autoPop} >
                  <View style={buttons.autoPopButtonContainer}>
                    <Icon name='magic' size={30} color='#fffdf9' style={{ marginTop: 5.5 }} />
                  </View>
                </TouchableOpacity>
                <TouchableOpacity
                  onPress={this.showQuestions} >
                  <View style={buttons.questionTab}>
                    <TabIcon showingQuestions={showingQuestions} />
                  </View>
                </TouchableOpacity>
              </View>
              <View style={containers.notesContainer}>
                {
                  showingQuestions &&
                  <View style={{ flex: 1 }}>
                    <FlatList
                      data={questions}
                      renderItem={({ item }) =>
                        <Question
                          id={item.id}
                          question={item.question}
                          save={this.saveQuestion}
                          delete={this.deleteQuestion}
                        />
                      }
                      keyExtractor={item => (item.id).toString()}
                      stickyHeaderIndices={[0]}
                      ListHeaderComponent={() =>
                        <View style={containers.questionsHeader}>
                          <Text style={text.questionHeaderText}>Questions</Text>
                          <View style={buttons.questionHeaderButtons}>
                            <TouchableOpacity
                              onPress={this.generateQuestions} >
                              <View style={buttons.questionGenButton}>
                                <Icon name='question-circle' size={25} />
                              </View>
                            </TouchableOpacity>
                            <TouchableOpacity>
                              <View style={buttons.notifButton}>
                                <Icon name='bell' size={20} />
                              </View>
                            </TouchableOpacity>
                          </View>
                        </View>
                      }
                      ListFooterComponent={() => <View />}
                      ListFooterComponentStyle={containers.emptyBlock}
                    />
                    {
                      showingAddQuestionButton &&
                      <TouchableOpacity
                        style={buttons.addQuestionButton}
                        onPress={this.addQuestion} >
                        <Icon name='plus' size={30} style={{ marginLeft: 16, }} />
                      </TouchableOpacity>
                    }
                  </View>
                }

                {
                  !showingQuestions &&
                  <TextInput
                    style={inputs.notesInput}
                    multiline={true}
                    onChangeText={(notesInput) => this.setState({ notesInput })}
                    value={notesInput}
                    placeholder={'Enter notes'}
                    textAlignVertical={'top'}
                  />
                }
              </View>
            </View>
          </View>
        </DismissKeyboard>
      </Modal>
    );
  }
};