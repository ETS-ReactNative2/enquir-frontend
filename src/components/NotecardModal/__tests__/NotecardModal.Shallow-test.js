import React from 'react';
import { TextInput } from 'react-native';
import { mount } from 'enzyme';

import NotecardModal from '../NotecardModal';

test('renders a NotecardModal using Enzyme 1', () => {
  const wrapper = mount(
    <NotecardModal
      ref={jest.fn()}
      show={false}
      onClose={jest.fn()}
      onSave={jest.fn()}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('renders a NotecardModal using Enzyme 2', () => {
  const wrapper = mount(
    <NotecardModal
      ref={jest.fn()}
      show={false}
      onClose={jest.fn()}
      onSave={jest.fn()}
    />
  );
  expect(wrapper).toMatchSnapshot();
});