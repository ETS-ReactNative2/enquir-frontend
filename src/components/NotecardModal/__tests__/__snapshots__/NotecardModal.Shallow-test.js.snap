// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a NotecardModal using Enzyme 1 1`] = `
<NotecardModal
  onClose={[MockFunction]}
  onSave={[MockFunction]}
  show={false}
/>
`;

exports[`renders a NotecardModal using Enzyme 2 1`] = `
<NotecardModal
  onClose={[MockFunction]}
  onSave={[MockFunction]}
  show={false}
/>
`;
